src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+
+                            break;
+                    }
+
+                    break;
+
+                case SyntaxKind.YieldReturnStatement:
+                    var memberDeclaration = semanticModel.GetEnclosingSymbol(forEachStatement.SpanStart, cancellationToken).DeclaringSyntaxReferences.Single().GetSyntax();
+                    var yieldStatements = memberDeclaration.DescendantNodes().OfType<YieldStatementSyntax>();
+
+                    if (forEachStatement.IsParentKind(SyntaxKind.Block) && forEachStatement.Parent.Parent == memberDeclaration)
+                    {
+                        var statementsOnBlockWithForEach = ((BlockSyntax)forEachStatement.Parent).Statements;
+                        var lastStatement = statementsOnBlockWithForEach.Last();
+                        if (yieldStatements.Count() == 1 && lastStatement == forEachStatement)
+                        {
+                            converter = new YieldReturnConverter(forEachStatement, convertingNodes, (YieldStatementSyntax)statementCannotBeConverted);
+                            return true;
+                        }
+
+                        // foreach()
+                        // {
+                        //   yield return ...;
+                        // }
+                        // yield break;
+                        // end of member
+                        if (yieldStatements.Count() == 2 &&
+                            lastStatement.Kind() == SyntaxKind.YieldBreakStatement &&
+                            statementsOnBlockWithForEach.ElementAt(statementsOnBlockWithForEach.Count - 2) == forEachStatement)
+                        {
+                            // This removes the yield break.
+                            converter = new YieldReturnConverter(forEachStatement, convertingNodes, (YieldStatementSyntax)statementCannotBeConverted, nodeToDelete: lastStatement);
+                            return true;
+                        }
+                    }
+
+                    break;
+            }
+
+            converter = default;
+            return false;
+        }
+
+        protected override SyntaxNode AddLinqUsing(SyntaxNode root)
+        {
+            const string linqNamespaceName = "System.Linq";
+            if (root is CompilationUnitSyntax compilationUnit)
+            {
+                var existingUsings = compilationUnit.Usings;
+                if (!existingUsings.Any(existingUsing => existingUsing.Name.ToString().Equals(linqNamespaceName)))
+                {
+                    var linqName = SyntaxFactory.ParseName(linqNamespaceName);
+                    return compilationUnit.AddUsings(SyntaxFactory.UsingDirective(linqName));
+                }
+            }
+
+            return root;
+        }
+
+        private static bool IsList(ITypeSymbol typeSymbol, SemanticModel semanticModel)
+            => Equals(typeSymbol.OriginalDefinition, semanticModel.Compilation.GetTypeByMetadataName(typeof(List<>).FullName));
```

Comment:
do you know that typeSymbol is non-null?  Consider:  typeSymbol?.OriginalDefinition.